src/legacy/core_plugins/interpreter/types/common.ts

```diff
@@ -0,0 +1,23 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/**
+ * Utility type: extracts returned type from a Promise.
+ */
+export type UnwrapPromise<T> = T extends Promise<infer P> ? P : T;
```

Comment:
My brain hurts!

Reply:
Clint was already using this in a generic somewhere, so I just extracted it as a utility. Unfortunately I can't take credit for authorship xD

Use case is when you have a type that may or may not be returned from a Promise -- you can use this and then you don't need to worry about whether the promise is there or not.

Reply:
Yeah, I understood it. It just took a bit of staring. I was just reminded of my experience learning Haskell a few years back. Hence, the aching brain. :)